osutil/inotify/inotify_linux.go

```diff
@@ -0,0 +1,319 @@
+		}
+	}
+
+	if !found {
+		w.watches[path] = &watch{wd: uint32(wd), flags: flags}
+		w.paths[wd] = path
+	}
+	w.mu.Unlock()
+	return nil
+}
+
+// Watch adds path to the watched file set, watching all events.
+func (w *Watcher) Watch(path string) error {
+	return w.AddWatch(path, InAllEvents)
+}
+
+// RemoveWatch removes path from the watched file set.
+func (w *Watcher) RemoveWatch(path string) error {
+	watch, ok := w.watches[path]
+	if !ok {
+		return fmt.Errorf("can't remove non-existent inotify watch for: %s", path)
+	}
+	success, errno := syscall.InotifyRmWatch(w.fd, watch.wd)
+	if success == -1 {
+		// when file descriptor or watch descriptor not found, InotifyRmWatch syscall return EINVAL error
+		// if return error, it may lead this path remain in watches and paths map, and no other event can trigger remove action.
+		if errno != syscall.EINVAL {
+			return os.NewSyscallError("inotify_rm_watch", errno)
+		}
+	}
+	delete(w.watches, path)
+	// Locking here to protect the read from paths in readEvents.
+	w.mu.Lock()
+	delete(w.paths, int(watch.wd))
+	w.mu.Unlock()
+	return nil
+}
+
+// readEvents reads from the inotify file descriptor, converts the
+// received events into Event objects and sends them via the Event channel
+func (w *Watcher) readEvents() {
+	var buf [syscall.SizeofInotifyEvent * 4096]byte
+	defer func() {
+		// Close w.Event first.
+		close(w.Event)
+		close(w.Error)
+	}()
+
+	for {
+		// wait until there are events from the kernel
+		n, err := w.notifyFile.Read(buf[:])
+
+		// If EOF is received
+		if n == 0 {
+			return
+		}
+		if n < 0 {
+			select {
+			case w.Error <- os.NewSyscallError("read", err):
+				continue
+			case <-w.done:
+				return
+			}
+		}
+		if n < syscall.SizeofInotifyEvent {
+			select {
+			case w.Error <- errors.New("inotify: short read in readEvents()"):
+				continue
+			case <-w.done:
+				return
+			}
+		}
+
+		var offset uint32
+		// We don't know how many events we just read into the buffer
+		// While the offset points to at least one whole event...
+		for offset <= uint32(n-syscall.SizeofInotifyEvent) {
+			// Point "raw" to the event in the buffer
+			raw := (*syscall.InotifyEvent)(unsafe.Pointer(&buf[offset]))
+			event := new(Event)
+			event.Mask = uint32(raw.Mask)
+			event.Cookie = uint32(raw.Cookie)
+			nameLen := uint32(raw.Len)
+			// The raw event received from the kernel contains only the file/folder name, but
+			// not the full path. To fix this, and fill the Event.Name field with the complete
+			// path of the file/folder, we retrieve it from the "paths" map using the Watch
+			// Descriptor (which is the one returned by InotifyAddWatch when we added the
+			// file/folder to being monitored).
+			w.mu.Lock()
+			name, ok := w.paths[int(raw.Wd)]
+			w.mu.Unlock()
+			if ok {
+				event.Name = name
+				if nameLen > 0 {
+					// Point "bytes" at the first byte of the filename
+					bytes := (*[syscall.PathMax]byte)(unsafe.Pointer(&buf[offset+syscall.SizeofInotifyEvent]))
```

Comment:
It's not clear why we're going through unsafe here, since we start with a []byte slice, and end with a byte slice.
```suggestion
					bytes := buf[offset+syscall.SizeofInotifyEvent:offset+syscall.SizeofInotifyEvent+nameLen]
```

Reply:
I don't know for sure, but checking this I suspect that it is concealing a bug: an inotify event has a minimum size of SizeofInotifyEvent, but that is if the nameLen field is zero. To detect if there is a complete event in the buffer, the code is just checking for that minimum length and then assumes that it contains a full event; but if the name falls outside the buffer, this is false...

I'll try to fix this.

Reply:
Ok, after searching info about inotify, it seems that it never sends "half events", so there is no risk of reading past the buffer.

Reply:
Changed, and also removed unneeded comparison.

Reply:
The "cast to a pointer to a very large array and then slice it to something smaller" is a hold-over from before the `unsafe.Slice` helper function was added to the language. But in this particular case, we don't even need `unsafe.Slice`, since we're just extracting a byte slice from within a larger byte slice.